File: packages/dawn/src/platform/darwin.cpp
#include "darwin.h"

#include <dlfcn.h>
#include <stdio.h>

template <typename T>
static T resolve(const char* name) {
  void* p = dlsym(RTLD_DEFAULT, name);
  T method = reinterpret_cast<T>(p);
  if (!method) {
    fprintf(stderr, "FFI stub: dlsym failed to resolve symbol: %s\n", name);
    return nullptr;
  }

  return method;
}

#define FFI_BIND(name, swift_name, ret, ...)         \
  using swift_name##_Fn = ret (*)(__VA_ARGS__);      \
  static swift_name##_Fn swift_name##_ptr = nullptr; \
  ret name(__VA_ARGS__)

#define FFI_RESOLVE(name) name##_ptr = resolve<name##_Fn>(#name)

FFI_BIND(flutter_texture_registry_get, DawnPluginGlobals_getTextureRegistry, flutter_texture_registry_t) {
  return DawnPluginGlobals_getTextureRegistry_ptr();
}

FFI_BIND(flutter_texture_registry_register, FlutterTextureRegistry_register, int64_t,
         flutter_texture_registry_t registry, flutter_texture_t texture) {
  return FlutterTextureRegistry_register_ptr(registry, texture);
}

FFI_BIND(flutter_texture_registry_texture_frame_available, FlutterTextureRegistry_textureFrameAvailable, void,
         flutter_texture_registry_t registry, int64_t textureId) {
  FlutterTextureRegistry_textureFrameAvailable_ptr(registry, textureId);
}

FFI_BIND(flutter_texture_registry_unregister_texture, FlutterTextureRegistry_unregisterTexture, void,
         flutter_texture_registry_t registry, int64_t textureId) {
  FlutterTextureRegistry_unregisterTexture_ptr(registry, textureId);
}

FFI_BIND(mtl_flutter_texture_create, MTLFlutterTexture_create, mtl_flutter_texture_t) {
  return MTLFlutterTexture_create_ptr();
}

FFI_BIND(mtl_flutter_texture_destroy, MTLFlutterTexture_destroy, void, mtl_flutter_texture_t texture) {
  MTLFlutterTexture_destroy_ptr(texture);
}

FFI_BIND(mtl_flutter_texture_update_buffer, MTLFlutterTexture_updateBuffer, void, mtl_flutter_texture_t texture,
         mtl_texture_t mtl_texture) {
  MTLFlutterTexture_updateBuffer_ptr(texture, mtl_texture);
}

void darwin_ffi_init() {
  FFI_RESOLVE(DawnPluginGlobals_getTextureRegistry);
  FFI_RESOLVE(FlutterTextureRegistry_register);
  FFI_RESOLVE(FlutterTextureRegistry_textureFrameAvailable);
  FFI_RESOLVE(FlutterTextureRegistry_unregisterTexture);
  FFI_RESOLVE(MTLFlutterTexture_create);
  FFI_RESOLVE(MTLFlutterTexture_destroy);
  FFI_RESOLVE(MTLFlutterTexture_updateBuffer);
}

File: packages/dawn/lib/src/platform/darwin.dart
import 'dart:ffi';

import 'gen/darwin_bindings.dart' as bindings;

class FlutterTextureRegistry {
  FlutterTextureRegistry._(this._ptr);

  static bool _initialized = false;
  static FlutterTextureRegistry? _instance;
  static FlutterTextureRegistry get instance {
    if (!_initialized) {
      bindings.darwin_ffi_init();
      _initialized = true;
    }

    return _instance ??= ._(bindings.flutter_texture_registry_get());
  }

  final bindings.flutter_texture_registry_t _ptr;

  int register(FlutterTexture texture) {
    final id = bindings.flutter_texture_registry_register(_ptr, texture._ptr.cast());
    texture._textureId = id;
    return id;
  }

  void textureFrameAvailable(int textureId) {
    bindings.flutter_texture_registry_texture_frame_available(_ptr, textureId);
  }

  void unregister(int textureId) {
    bindings.flutter_texture_registry_unregister_texture(_ptr, textureId);
  }
}

abstract class FlutterTexture {
  FlutterTexture._(this._ptr);

  final Pointer _ptr;
  int? _textureId;

  int? get textureId => _textureId;
  bool get isRegistered => _textureId != null;

  int register() {
    assert(_textureId == null);
    return FlutterTextureRegistry.instance.register(this);
  }

  void frameAvailable() {
    assert(_textureId != null);
    FlutterTextureRegistry.instance.textureFrameAvailable(_textureId!);
  }

  void unregister() {
    assert(_textureId != null);
    FlutterTextureRegistry.instance.unregister(_textureId!);
  }
}

class MTLFlutterTexture extends FlutterTexture {
  MTLFlutterTexture._(super._ptr) : super._() {
    _finalizer.attach(this, (_ptr.cast(), null), detach: this);
  }

  factory MTLFlutterTexture.create() {
    final ptr = bindings.mtl_flutter_texture_create();
    return ._(ptr);
  }

  static final _finalizer = Finalizer<(bindings.mtl_flutter_texture_t, int?)>((data) {
    final (ptr, id) = data;
    _dispose(ptr, id);
  });

  static void _dispose(bindings.mtl_flutter_texture_t ptr, int? textureId) {
    if (textureId != null) {
      final registry = bindings.flutter_texture_registry_get();
      bindings.flutter_texture_registry_unregister_texture(registry, textureId);
    }

    bindings.mtl_flutter_texture_destroy(ptr);
  }

  void _refreshFinalizer() {
    _finalizer.detach(this);
    _finalizer.attach(this, (_ptr.cast(), _textureId), detach: this);
  }

  @override
  int register() {
    final id = super.register();
    _refreshFinalizer();
    return id;
  }

  void updateBuffer(Pointer<Void> mtlTexture) {
    bindings.mtl_flutter_texture_update_buffer(_ptr.cast(), mtlTexture.cast());
  }

  @override
  void unregister() {
    super.unregister();
    _refreshFinalizer();
  }

  void dispose() {
    _finalizer.detach(this);
    _dispose(_ptr.cast(), _textureId);
  }
}
